antlr/src/main/antlr4/com/baeldung/antlr/ServerLog.g4

```diff
@@ -0,0 +1,34 @@
+grammar ServerLog;
+
+log : logEntry+ EOF ;
+logEntry : serverTime WS level WS message NEWLINE;
+serverTime : date WS dateTime;
+message : (WORD | WS)+ ;
+level : ('DEBUG' | 'INFO' | 'ERROR');
+date : NUMBER DASH month DASH NUMBER;
```

Comment:
Some of these lines have a space before the semi-colon. Would recommend being consistent either way.

Also, see my comments in JIRA about a way to maybe simplify this further. I think it is okay, but let's just remember to only be as complex as we have to, especially when building a grammar file, while important, isn't the main focus of the article.